### PR TITLE
Adds support for Promises returned in config, fixes #1503

### LINKED ISSUE #1503: Return promise from config file
My config is using async operations to read the file system and exports a promise to be fulfilled with the config instead of the config itself but looks like rollup is not prepared for that:

> 🚨   Config file must export an options object
> https://github.com/rollup/rollup/wiki/Command-Line-Interface#using-a-config-file

Would it be hard to support promises on config file output?

### PR DESCRIPTION
**Pull request by ankeetmaini**
This PR is raised against #1503. 

- to account for cases where the config file actually does some async work to get options
- added a test too

### CODE REVIEW

**Comment by ankeetmaini** [comment]
cc: @Rich-Harris, @lukastaegert let me know if you still want this feature. Thanks :)

**Comment by lukastaegert** [comment]
@ankeetmaini Yes, I find this VERY interesting! During JS Kongress this week I learned a lot about people using rollup in very creative ways to build very complex pipelines and I am sure those people would love this functionality.
My review pipeline has just grown a little long so stay tuned!

**Comment by ankeetmaini** [comment]
@lukastaegert thanks so much! I'm interested in contributing to Rollup, is there an architecture note on how to understand the codebase?

**Comment by lukastaegert** [comment]
Unfortunately not yet and, to be honest, things are a little in flux especially regarding the core tree-shaking algorithm. I hope to be able to write something up once things have settled down a little but probably not too soon.

👍🎉 for people who want to contribute!

If you want some ideas where to go next:
* Things that improve the user experience are always heartily welcomed! I do not have a lot of time to look into these things and honestly, this is something that others are probably better suited to do. Your PR goes in this direction but we also need people to take care the documentation stays up to date with these enhancements.
* Also to my knowledge, there are inconsistencies between the CLI interface and the config options that should be resolved at some point.
* I hear there are some issues related to caching that I do not expect to find any time to look at soon.
* Or anything else from the issues list 😁 Maybe make a comment if start to work at a PR for an issue. That way, people are aware of this and might give you input and it would also prevent work being done twice.

File: test/cli/samples/config-promise/_expected.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+(function (global, factory) {
```
Comment (lukastaegert) [inline test/cli/samples/config-promise/_expected.js]: Including this file does not have an effect as it is only checked if you do NOT set `execute: true` in `_config.js`. You should remove it as it is rather confusing (and while you are at it, remove it from the `cli/config` test as well 😜)

Comment (ankeetmaini) [reply test/cli/samples/config-promise/_expected.js]: You got it!

**Comment by lukastaegert** [comment]
BTW we should document this on the web site. If someone were to prepare a PR for https://github.com/rollup/rollupjs.org that describes this feature and at the same time also addresses rollup/rollupjs.org#69, that would be awesome!

**Comment by ankeetmaini** [comment]
I'll take a look!

**Comment by lukastaegert** [comment]
I have started to assemble a branch for the next release but it will take another few days until I release it because I there are a few more PRs I want to have in there which are not ready yet.

**Pull request merged**